File: src/asrai_mcp/tools.py
import asyncio
import contextvars
import json
import os
from eth_account import Account
from x402.clients.httpx import x402HttpxClient

BASE_URL = "https://x402.asrai.me"
X402_HEADERS = {"x-coinbase-402": "true", "x-payment-token": "usdc"}

# Per-connection state via ContextVar — works for both stdio (single-user)
# and SSE (multi-user) modes without any code changes to tools.
_current_account: contextvars.ContextVar = contextvars.ContextVar("current_account", default=None)
_connection_spend: contextvars.ContextVar = contextvars.ContextVar("connection_spend", default=0.0)


def _check_spend(amount: float):
    max_s = float(os.environ.get("ASRAI_MAX_SPEND", "2.0"))
    current = _connection_spend.get(0.0)
    new_total = current + amount
    _connection_spend.set(new_total)
    if new_total > max_s:
        raise ValueError(
            f"Session spend limit of ${max_s} USDC reached. "
            f"Set ASRAI_MAX_SPEND env var to increase."
        )


def _get_account():
    # SSE multi-user mode: account set per-connection by sse_server.py
    ctx_account = _current_account.get(None)
    if ctx_account:
        return ctx_account
    # Stdio mode (Claude Desktop / openclaw): read from env as before
    key = os.environ.get("PRIVATE_KEY")
    if not key:
        raise ValueError("PRIVATE_KEY environment variable is required")
    return Account.from_key(key)


async def _get(path: str) -> dict | list | str:
    _check_spend(0.001)
    async with x402HttpxClient(account=_get_account(), base_url=BASE_URL, timeout=30.0) as client:
        r = await client.get(path, headers=X402_HEADERS)
        raw = (await r.aread()).decode()
        try:
            return json.loads(raw)
        except Exception:
            return raw


async def _post(path: str, body: dict) -> dict | list | str:
    _check_spend(0.001)
    async with x402HttpxClient(account=_get_account(), base_url=BASE_URL, timeout=30.0) as client:
        r = await client.post(path, json=body, headers=X402_HEADERS)
        raw = (await r.aread()).decode()
        try:
            return json.loads(raw)
        except Exception:
            return raw


async def _gather(*paths: str) -> dict:
    # Sequential to avoid x402 parallel payment conflicts from same wallet
    results = {}
    for path in paths:
        try:
            results[path] = await _get(path)
        except Exception as e:
            results[path] = str(e)
    return results


# ---------------------------------------------------------------------------
# Tool handlers
# ---------------------------------------------------------------------------

async def market_overview() -> str:
    data = await _gather(
        "/api/trending/",
        "/api/gainers-losers/",
        "/api/rsi/",
        "/api/top-bottom/",
    )
    return json.dumps(data, indent=2)


async def technical_analysis(symbol: str, timeframe: str = "1D") -> str:
    s = symbol.lower().rstrip("usdt")
    data = await _gather(
        f"/api/signal/{s}usdt/{timeframe}",
        f"/api/alsat/{s}usdt/{timeframe}",
        f"/api/superalsat/{s}usdt",
        f"/api/psar/{s}usdt/{timeframe}",
        f"/api/macd-dema/{s}usdt/{timeframe}",
        f"/api/alphatrend/{s}usdt/{timeframe}",
        f"/api/td/{s}usdt/{timeframe}",
    )
    return json.dumps(data, indent=2)


async def sentiment() -> str:
    data = await _gather(
        "/api/cbbi/",
        "/api/cmc-sentiment/",
        "/api/cmcai/",
    )
    return json.dumps(data, indent=2)


async def forecast(symbol: str) -> str:
    s = symbol.lower().rstrip("usdt")
    data = await _get(f"/api/forecasting/{s}usdt")
    return json.dumps(data, indent=2)


async def screener(screener_type: str) -> str:
    valid = [
        "ichimoku-trend", "sar-coins", "macd-coins", "emacross",
        "techrating", "vwap", "volume", "highvolumelowcap",
        "bounce-dip", "galaxyscore", "socialdominance", "late-unlocked-coins",
        "ath", "rsi", "rsi-heatmap", "ao",
    ]
    if screener_type not in valid:
        return json.dumps({"error": f"Invalid screener. Choose from: {valid}"})
    data = await _get(f"/api/{screener_type}/")
    return json.dumps(data, indent=2)


async def smart_money(symbol: str, timeframe: str = "1D") -> str:
    s = symbol.lower().rstrip("usdt")
    data = await _gather(
        f"/api/smartmoney/{s}usdt/{timeframe}",
        f"/api/support-resistance/{s}usdt/{timeframe}",
    )
    return json.dumps(data, indent=2)


async def elliott_wave(symbol: str, timeframe: str = "1D") -> str:
    s = symbol.lower().rstrip("usdt")
    data = await _get(f"/api/ew/{s}usdt/{timeframe}")
    return json.dumps(data, indent=2)


async def ichimoku(symbol: str, timeframe: str = "1D") -> str:
    s = symbol.lower().rstrip("usdt")
    data = await _get(f"/api/ichimoku/{s}usdt/{timeframe}")
    return json.dumps(data, indent=2)


async def cashflow(mode: str, symbol: str = "") -> str:
    if mode == "market":
        data = await _get("/api/cashflow/market")
    elif mode == "coin" and symbol:
        data = await _get(f"/api/cashflow/coin/{symbol.lower()}")
    elif mode == "group" and symbol:
        data = await _get(f"/api/cashflow/group/{symbol.lower()}")
    else:
        return json.dumps({"error": "mode must be 'market', 'coin', or 'group'. coin/group require symbol."})
    return json.dumps(data, indent=2)


async def coin_info(symbol: str) -> str:
    s = symbol.lower()
    data = await _gather(f"/api/coinstats/{s}", f"/api/info/{s}", f"/api/price/{s}", f"/api/tags/{s}")
    return json.dumps(data, indent=2)


async def dexscreener(contract_address: str, chain: str = "") -> str:
    if chain:
        data = await _get(f"/api/dexscreener/{chain}/{contract_address}")
    else:
        data = await _get(f"/api/dexscreener/{contract_address}")
    return json.dumps(data, indent=2)


async def chain_tokens(chain: str, max_mcap: str) -> str:
    data = await _get(f"/api/chain/{chain}/{max_mcap}")
    return json.dumps(data, indent=2)


async def portfolio(symbol: str = "") -> str:
    path = f"/api/portfolio/{symbol.lower()}" if symbol else "/api/portfolio/"
    data = await _get(path)
    return json.dumps(data, indent=2)


async def channel_summary() -> str:
    data = await _get("/api/channel-summary/")
    return json.dumps(data, indent=2)


async def ask_ai(question: str) -> str:
    data = await _post("/ai", {"message": question})
    return json.dumps(data, indent=2)


from asrai_mcp.indicators_guide import indicator_guide  # noqa: F401

File: src/tools.js
/**
 * Asrai API calls via x402 payment protocol.
 * Each call costs $0.001 USDC from the user's wallet on Base mainnet.
 *
 * Supports two modes:
 *  - stdio (Claude Desktop): reads PRIVATE_KEY from env
 *  - SSE (Docker/n8n): key injected per-connection via AsyncLocalStorage
 */

import { AsyncLocalStorage } from "node:async_hooks";
import { wrapFetchWithPayment, x402Client } from "@x402/fetch";
import { registerExactEvmScheme } from "@x402/evm/exact/client";
import { privateKeyToAccount } from "viem/accounts";

const BASE_URL = "https://x402.asrai.me";
const X402_HEADERS = { "x-coinbase-402": "true", "x-payment-token": "usdc" };

// Per-connection state — works for both stdio and multi-user SSE
// In stdio mode: nothing set here, falls back to PRIVATE_KEY env var
// In SSE mode: each connection runs inside connectionStorage.run({ key, spend: 0 }, ...)
export const connectionStorage = new AsyncLocalStorage();

function getMaxSpend() {
  return parseFloat(process.env.ASRAI_MAX_SPEND ?? "2.0");
}

function checkSpend(amount) {
  const store = connectionStorage.getStore();
  if (store) {
    store.spend = (store.spend ?? 0) + amount;
    if (store.spend > getMaxSpend()) {
      throw new Error(
        `Session spend limit of $${getMaxSpend()} USDC reached. ` +
        `Set ASRAI_MAX_SPEND env var to increase.`
      );
    }
  }
}

function buildFetch() {
  // SSE mode: key from per-connection storage
  // stdio mode: key from env var
  const store = connectionStorage.getStore();
  const key = store?.key ?? process.env.PRIVATE_KEY;
  if (!key) throw new Error("PRIVATE_KEY environment variable is required");

  const signer = privateKeyToAccount(key);
  const client = new x402Client();
  registerExactEvmScheme(client, { signer });
  return wrapFetchWithPayment(fetch, client);
}

async function _get(path) {
  checkSpend(0.001);
  const fetchWithPayment = buildFetch();
  const res = await fetchWithPayment(`${BASE_URL}${path}`, {
    headers: X402_HEADERS,
    signal: AbortSignal.timeout(90_000),
  });
  const text = await res.text();
  try { return JSON.parse(text); } catch { return text; }
}

async function _post(path, body) {
  checkSpend(0.001);
  const fetchWithPayment = buildFetch();
  const res = await fetchWithPayment(`${BASE_URL}${path}`, {
    method: "POST",
    headers: { ...X402_HEADERS, "Content-Type": "application/json" },
    body: JSON.stringify(body),
    signal: AbortSignal.timeout(90_000),
  });
  const text = await res.text();
  try { return JSON.parse(text); } catch { return text; }
}

async function _gather(...paths) {
  // Sequential — parallel x402 payments from same wallet can conflict
  const results = {};
  for (const path of paths) {
    try { results[path] = await _get(path); }
    catch (e) { results[path] = String(e); }
  }
  return results;
}

// ── Symbol normalisation ────────────────────────────────────────────────────

function sym(symbol) {
  return symbol.toLowerCase().replace(/usdt$/, "");
}

// ── Tool handlers ───────────────────────────────────────────────────────────

export async function market_overview() {
  return JSON.stringify(await _gather(
    "/api/trending/",
    "/api/gainers-losers/",
    "/api/rsi/",
    "/api/top-bottom/",
  ), null, 2);
}

export async function technical_analysis(symbol, timeframe = "1D") {
  const s = sym(symbol);
  return JSON.stringify(await _gather(
    `/api/signal/${s}usdt/${timeframe}`,
    `/api/alsat/${s}usdt/${timeframe}`,
    `/api/superalsat/${s}usdt`,
    `/api/psar/${s}usdt/${timeframe}`,
    `/api/macd-dema/${s}usdt/${timeframe}`,
    `/api/alphatrend/${s}usdt/${timeframe}`,
    `/api/td/${s}usdt/${timeframe}`,
  ), null, 2);
}

export async function sentiment() {
  return JSON.stringify(await _gather(
    "/api/cbbi/",
    "/api/cmc-sentiment/",
    "/api/cmcai/",
  ), null, 2);
}

export async function forecast(symbol) {
  return JSON.stringify(await _get(`/api/forecasting/${sym(symbol)}usdt`), null, 2);
}

export async function screener(screener_type) {
  const valid = [
    "ichimoku-trend", "sar-coins", "macd-coins", "emacross",
    "techrating", "vwap", "volume", "highvolumelowcap",
    "bounce-dip", "galaxyscore", "socialdominance", "late-unlocked-coins",
    "ath", "rsi", "rsi-heatmap", "ao",
  ];
  if (!valid.includes(screener_type)) {
    return JSON.stringify({ error: `Invalid screener. Choose from: ${valid.join(", ")}` });
  }
  return JSON.stringify(await _get(`/api/${screener_type}/`), null, 2);
}

export async function smart_money(symbol, timeframe = "1D") {
  const s = sym(symbol);
  return JSON.stringify(await _gather(
    `/api/smartmoney/${s}usdt/${timeframe}`,
    `/api/support-resistance/${s}usdt/${timeframe}`,
  ), null, 2);
}

export async function elliott_wave(symbol, timeframe = "1D") {
  return JSON.stringify(await _get(`/api/ew/${sym(symbol)}usdt/${timeframe}`), null, 2);
}

export async function ichimoku(symbol, timeframe = "1D") {
  return JSON.stringify(await _get(`/api/ichimoku/${sym(symbol)}usdt/${timeframe}`), null, 2);
}

export async function cashflow(mode, symbol = "") {
  if (mode === "market") {
    return JSON.stringify(await _get("/api/cashflow/market"), null, 2);
  }
  if ((mode === "coin" || mode === "group") && symbol) {
    return JSON.stringify(await _get(`/api/cashflow/${mode}/${symbol.toLowerCase()}`), null, 2);
  }
  return JSON.stringify({ error: "mode must be 'market', 'coin', or 'group'. coin/group require symbol." });
}

export async function coin_info(symbol) {
  const s = symbol.toLowerCase();
  return JSON.stringify(await _gather(
    `/api/coinstats/${s}`,
    `/api/info/${s}`,
    `/api/price/${s}`,
    `/api/tags/${s}`,
  ), null, 2);
}

export async function dexscreener(contract_address, chain = "") {
  const path = chain
    ? `/api/dexscreener/${chain}/${contract_address}`
    : `/api/dexscreener/${contract_address}`;
  return JSON.stringify(await _get(path), null, 2);
}

export async function chain_tokens(chain, max_mcap) {
  return JSON.stringify(await _get(`/api/chain/${chain}/${max_mcap}`), null, 2);
}

export async function portfolio(symbol = "") {
  const path = symbol ? `/api/portfolio/${symbol.toLowerCase()}` : "/api/portfolio/";
  return JSON.stringify(await _get(path), null, 2);
}

export async function channel_summary() {
  return JSON.stringify(await _get("/api/channel-summary/"), null, 2);
}

export async function ask_ai(question) {
  return JSON.stringify(await _post("/ai", { message: question }), null, 2);
}
